docs/src/rfcs/002-fearless-concurrency.md

```diff
@@ -0,0 +1,220 @@
+
+For now these interfaces have two functions - but will likely change to include additional scopes (e.g. subgroup) and operations (e.g. reduction, expansion) in future iterations of the proposal:
+
+```rust
+impl<T> MutDeviceStorage {
+    fn map_workgroup<F, E>(&mut self, f: F) where F: FnMut(E) -> E
+    {
+        unsafe {
+            spv::op_memory_barrier(Workgroup, Acquire);
+            let value: E = self.data.load_indexed_from_workgroup(invocation_index());  // OpLoad with NonPrivate|MakeVisible semantics and Workgroup scope.
+            let result = f(value);
+            self.data.store_indexed_to_workgroup(invocation_index(), result);          // OpStore with NonPrivate|MakeAvailable semantics and Workgroup scope
+            spv::op_memory_barrier(Workgroup, Release);
+        }
+    }
+    
+    fn map_workgroup_joined<F, E, const RATE: usize>(&mut self, f: F) where F: FnMut([E; RATE]) -> [E; RATE]
+    {
+        unsafe {
+            spv::op_control_barrier(Workgroup, Workgroup, Acquire);
+            if (invocation_index % RATE == 0) {
+                let value: [E; RATE] = self.data.load_array_indexed_from_workgroup(invocation_index());  // OpLoad with NonPrivate|MakeVisible semantics and Workgroup scope.
+                let result = f(value);
+                self.data.store_array_indexed_to_workgroup(invocation_index(), result);                  // OpStore with NonPrivate|MakeAvailable semantics and Workgroup scope
+            }
+            spv::op_control_barrier(Workgroup, Workgroup, Release);
+        }
+    }
+}
+```
+
+`map_workgroup` gives you a simple 1:1 interface for mutating values in a buffer, allowing you to mutate invocation-local data. It does so safely, with barriers and visibility automatically executed to provide safety.
+
+`map_workgroup_joined` is similar, but allows a shader to access data shared by multiple threads, so that it can be shuffled or combined across threads. The number of threads that data is shared between is fixed by the `RATE` parameter. This function must only be called in workgroup-uniform control flow, and `RATE` must be a factor of the workgroup size.
+
+The `MutWorkgroupStorage` implementation is identical; it just operates on Workgroup storage instead.
+
+The load and store variants used on the underlying storage here are invented to simplify the interface; what these ultimately look like is contingent on the resource binding proposal, which is still in flight.
+The spir-v op codes are similarly somewhat invented, and will also be adjusted to mirror the final syntax for assembly here (or some friendlier wrapper).
+These are provided primarily for illustrative purposes.
+
+
+### Example
+
+Bitonic sort example (unrolled, incomplete...):
+
+```rust
+fn bitonic_sort_workgroup(buffer: MutDeviceStorage<u32>) {
+    buffer.map_workgroup_joined<2>(|array| {
+        [
+            max(array[0],array[1]),
+            min(array[0],array[1]),
+        ]
+    });
```

Comment:
Can't this length be inferred by the compiler? I would assume the below would work. Also the syntax isn't `<2>` it's `::<2>`
```suggestion
    buffer.map_workgroup_joined(|array| {
        [
            max(array[0],array[1]),
            min(array[0],array[1]),
        ]
    });
```

Reply:
Hm yep you're right. I'll fix this.